pkg/aws/cloud.go

```diff
@@ -84,54 +92,67 @@ func NewCloud(cfg CloudConfig, metricsRegisterer prometheus.Registerer, logger l
 
 		if region == "" {
 			err := (error)(nil)
-			region, err = metadata.Region()
+			region, err = ec2Metadata.Region()
 			if err != nil {
 				return nil, errors.Wrap(err, "failed to introspect region from EC2Metadata, specify --aws-region instead if EC2Metadata is unavailable")
 			}
 		}
 		cfg.Region = region
 	}
-	awsCFG := aws.NewConfig().WithRegion(cfg.Region).WithSTSRegionalEndpoint(endpoints.RegionalSTSEndpoint).WithMaxRetries(cfg.MaxRetries).WithEndpointResolver(endpointsResolver)
-	opts = session.Options{}
-	opts.Config.MergeIn(awsCFG)
-	if !hasIPv4 {
-		opts.EC2IMDSEndpointMode = endpoints.EC2IMDSEndpointModeStateIPv6
-	}
-	sess := session.Must(session.NewSessionWithOptions(opts))
-	injectUserAgent(&sess.Handlers)
+	awsConfig, err := config.LoadDefaultConfig(context.TODO(),
+		config.WithRegion(cfg.Region),
+		config.WithRetryer(func() aws.Retryer {
+			return retry.NewStandard(func(o *retry.StandardOptions) {
+				o.RateLimiter = ratelimit.None
+				o.MaxAttempts = cfg.MaxRetries
+			})
+		}),
+		config.WithEC2IMDSEndpointMode(ec2IMDSEndpointMode),
+		config.WithAPIOptions([]func(stack *smithymiddleware.Stack) error{
+			awsmiddleware.AddUserAgentKeyValue(userAgent, version.GitVersion),
+		}),
+	)
 
 	if cfg.ThrottleConfig != nil {
 		throttler := throttle.NewThrottler(cfg.ThrottleConfig)
-		throttler.InjectHandlers(&sess.Handlers)
+		awsConfig.APIOptions = append(awsConfig.APIOptions, func(stack *smithymiddleware.Stack) error {
+			return throttle.WithSDKRequestThrottleMiddleware(throttler)(stack)
+		})
 	}
+
 	if metricsRegisterer != nil {
 		metricsCollector, err := metrics.NewCollector(metricsRegisterer)
 		if err != nil {
 			return nil, errors.Wrapf(err, "failed to initialize sdk metrics collector")
 		}
-		metricsCollector.InjectHandlers(&sess.Handlers)
+		awsConfig.APIOptions = append(awsConfig.APIOptions, func(stack *smithymiddleware.Stack) error {
+			return metrics.WithSDKCallMetricCollector(metricsCollector)(stack)
```

Comment:
i'd prefer to have a single method like `WithSDKMetricsCollector` instead of two. If we want to support optional collect either call/request metrics only, we can pass the setting into WithSDKMetricsCollector.

Reply:
Makes sense. I have fixed this in recent commit